File: scripts/perl/diffAtlas_DE_limma.pl
#!/usr/bin/perl
#
# Microarray differential expression analysis using limma for Differential Atlas.
# Read assay group and contrast definitions;
# Compute statistics for differential expression for each contrast using limma
# R package;
# Create "MvA" plots for each contrast using ggplot2 R package.
#
# To run:
# 	./diffAtlas_DE_limma.pl <*-contrasts.tsv> <normalized expressions file> </path/to/diffAtlas_DE_limma.R> </path/to/diffAtlas_mvaPlot.R>

use strict;
use warnings;

use File::Basename;

$| = 1;

# Commandline arguments are:
# 	- $defsFile : file with assignment of assay accessions to assay groups and assay groups to contrasts.
# 	- $normExpr : file with normalized, summarized expressions.
# 	- $limmaScript : filename of R script for differential expression analysis using limma.
# 	- $mvaScript : filename of R script for creating "MvA" plot with ggplot2.
$_ = shift for my ($defsFile, $normExpr, $limmaScript, $mvaScript);

my @accessions = split "_", basename($defsFile);
my $exptAcc = $accessions[0];
(my $arrayDesAcc = $accessions[1]) =~ s/-contrasts\.txt//;

unless($exptAcc && $arrayDesAcc) { die "Error: Couldn't find experiment and array design accessions in filename: $defsFile\n"; }

print "\nExperiment: $exptAcc\nArray design: $arrayDesAcc\n";


# Read data from contrast definitions file into hash:
# contrastHash->{ "contrast 1" }->{ "reference" } = [ "wildType1", "wildType2", "wildType3" ]
#                               ->{ "test" } = [ "Xmutant1", "Xmutant2", "Xmutant3" ]
#                               ->{ "AGnums" } = "2_1"
#             ->{ "contrast 2" }->{ "reference" } = [ "wildType1", "wildType2", "wildType3" ]
#                               ->{ "test" } = [ "Ymutant1", "Ymutant2", "Ymutant3" ]
#                               ->{ "AGnums" } = "2_3"
my $contrastHash = readContrasts($defsFile);
# Check we've got things in $contrastHash.
unless(%{ $contrastHash }) { die "Error: No contrasts found in $defsFile\n"; }


# Run differential expression analysis and make MvA plots.
# $limmaResHash->{ "designElement1" }->{ "contrast 1" } = [ "p-value", "t-stat", "logFC" ]
# 		                                 ->{ "contrast 2" } = [ "p-value", "t-stat", "logFC" ]
# 		           ->{ "designElement2" }->{ "contrast 1" } = [ "p-value", "t-stat", "logFC" ]
#                                        ->{ "contrast 2" } = [ "p-value", "t-stat", "logFC" ]
my $limmaResHash = getDEresults($contrastHash, $limmaScript, $normExpr, $exptAcc, $arrayDesAcc);


# Write all differential expression stats to a file.
writeLimmaResults($exptAcc, $arrayDesAcc, $limmaResHash);

# end
#####





###############
# subroutines #
###############

# readContrasts
#  - read the assay group and contrast definitions file and return a hash with
#  this data in.
sub readContrasts {

	my $defsFile = shift;
	my $contrastHash = {};

	# a temporary place to put assay groups
	# $agHash->{ "1" } = [ "wildType1", "wildType2", "wildType3" ]
	#        ->{ "2" } = [ "Xmutant1", "Xmutant2", "Xmutant3" ]
	#        ->{ "3" } = [ "Ymutant1", "Ymutant2", "Ymutant3" ]
	my $agHash = {};


	open(DEFSFILE, $defsFile) or die("Can't open $defsFile: $!\n");
	while(defined(my $line = <DEFSFILE>)) {
		
		unless($line =~ /^#/) {
			
			chomp $line;

			if($line =~ /^(\d+)=(.*)/) {

				my @ags = split "@@", $2;
				
				$agHash->{$1} = \@ags;
			}
			elsif($line =~ /=\d+ \d+/) {
				
				# given ordering of the contrasts file we should have got the assay
				# groups information by the time we get to the contrast
				# definitions. If not, die.
				unless(%{ $agHash }) { die "No assay groups found!\n"; }
				
				my @lineSplit = split "=", $line;
				my $contrastName = $lineSplit[0];
				my ($refAG, $testAG) = split " ", $lineSplit[1];

				$contrastHash->{ $contrastName }->{ "reference" } = $agHash->{ $refAG };
				$contrastHash->{ $contrastName }->{ "test" } = $agHash->{ $testAG };
				# This is for the plot filename:
				$contrastHash->{ $contrastName }->{ "AGnums" } = $refAG."_".$testAG;
			}
		}
	}
	
	my $numAGs = keys %{ $agHash };
	my $numContrasts = keys %{ $contrastHash };
	print "$numAGs assay groups and $numContrasts contrasts.\n\n";

	return $contrastHash;
}





# getDEresults
#  - Run limma script for each contrast and build a hash of results of all
#  contrasts:
# $limmaResultsHash->{ "designElement1" }->{ "contrast 1" } = [ "p-value", "t-stat", "logFC" ]
# 		                                 ->{ "contrast 2" } = [ "p-value", "t-stat", "logFC" ]
# 		           ->{ "designElement2" }->{ "contrast 1" } = [ "p-value", "t-stat", "logFC" ]
#                                        ->{ "contrast 2" } = [ "p-value", "t-stat", "logFC" ]
#
#  - Make an MvA plot for each contrast.
sub getDEresults {

	$_ = shift for my ($contrastHash, $limmaScript, $normExpr, $exptAcc, $arrayDesAcc);

	my $limmaResHash = {};
	
	# Filenames for temporary storage of limma results
	my $limmaResTempFile = ".limma_res.txt";
	my $plotDataTempFile = ".plotData.txt";
	
	# Do differential expression stats for each contrast
	foreach my $contrastName (keys %{ $contrastHash }) {

		# Reference and test assay accessions for limma script.
		my $refAssays = join ",", @{ $contrastHash->{ $contrastName }->{ "reference" }};
		my $testAssays = join ",", @{ $contrastHash->{ $contrastName }->{ "test" }};

		print "Computing differential expression statistics for contrast \"$contrastName\"...";
		
		# Run limma script.
		my $R_limmaOutput = `Rscript $limmaScript $normExpr $refAssays $testAssays $limmaResTempFile $plotDataTempFile 2>&1`;
		
		# Check for errors.
		if($R_limmaOutput =~ /error/i) {
			
			# Can't continue without results from limma so may as well quit.
			die("\nError during differential expression analysis, outout from R below.\n------------\n$R_limmaOutput\n------------\n");
		
		} else {
			print "done\n";
		}


		# Add results to hash.
		open(LIMMARES, "<$limmaResTempFile") or die("Can't open file $limmaResTempFile: $!\n");
		while(defined(my $line = <LIMMARES>)) {
			
			# potentially don't need headers in the limma results file?
			unless($line =~ /^designElements/) {
				
				chomp $line;
				
				# Split on tabs
				my @lineSplit = split "\t", $line;
				
				# Design element (probeset) ID is the first element, take it off.
				my $designElement = shift @lineSplit;
				
				# Add array [p-value, t-statistic, logFoldChange] to hash for this
				# design element for this contrast.
				$limmaResHash->{ $designElement }->{ $contrastName } = \@lineSplit;
			}
		}
		close(LIMMARES);
		
		
		# Filename for MvA plot
		my $plotFile = $exptAcc."-".$arrayDesAcc."-".$contrastHash->{ $contrastName }->{ "AGnums" }."-mvaPlot.png";
		
		print "Making MvA plot...";
		# Create MvA plot with MvA plot script
		my $R_mvaOutput = `Rscript $mvaScript $plotDataTempFile \"$contrastName\" $plotFile microarray 2>&1`;
		
		if($R_mvaOutput =~ /error/i) {

			# Report the error but don't worry about dying as we can live
			# without the MvA plot.
			print "\nError creating MvA plot, output from R below.\n------------\n$R_mvaOutput\n------------\n";

		} else {
			print "done.\n";
		}

		# clean up
		`rm $limmaResTempFile`;
		`rm $plotDataTempFile`;
		# mysterious Rplots.pdf
		`rm Rplots.pdf`;
	}

	print "\nStatistics for all contrasts computed.\n";
	return $limmaResHash;
}





# writeLimmaResults
#  - Write results from limma to tab-delimited text file.
sub writeLimmaResults {

	$_ = shift for my ($exptAcc, $arrayDesAcc, $limmaResHash);
	
	# Make filename
	my $limmaResFile = $exptAcc."-".$arrayDesAcc."-analytics.tsv";

	print "Writing results to $limmaResFile\n";

	open(RESFILE, ">$limmaResFile") or die("Can't open $limmaResFile: $!\n");

	# write column headers
	printf(RESFILE "Design Elements");
	foreach my $contrastName (keys %{ $contrastHash }) {
		printf(RESFILE "\t$contrastName.p-value\t$contrastName.t-statistic\t$contrastName.log2foldchange");
	}

	# write statistics
	foreach my $designElement (keys %{ $limmaResHash }) {

		printf(RESFILE "\n$designElement");
		
		foreach my $contrastName (keys %{ $limmaResHash->{ $designElement }}) {

			my $statsString = join "\t", @{ $limmaResHash->{ $designElement }->{ $contrastName }};

			printf(RESFILE "\t".$statsString);
		}
	}

	close(RESFILE);
}

File: scripts/R/diffAtlas_DE_limma.R
#!/ebi/microarray/home/biocep/local/lib64/R/bin/Rscript

# diffAtlas_DE_limma.R
# Microarray differential expression statistics computation for the
# Differential Atlas.


# diffAtlas_DE_limma()
# - Differential expression analysis (2-group comparison) using limma.
# Arguments:
# 	- normExprsFile <- matrix of normalized and summarized expression values.
#	- refAssays <- comma-separated list of assay accessions in reference assay group.
#	- testAssays <- comma-separated list of assay accessions in test assay group.
#	- resFile <- filename for results.
#	- plotDataFile <- filename for data for MvA plot.
#	- aValuesFile <- filename for matrix of A-values; only needed for 2-colour designs. Default is NULL.
diffAtlas_DE_limma <<- function(normExprsFile, refAssays, testAssays, resFile, plotDataFile, aValuesFile = NULL) {

	e <- try({
		
		# load limma.
		library(limma)
		
		# Read expressions (M-values/log-fold-changes for 2-colour).
		print(paste("Reading", normExprsFile))
		# R's reserved characters are changed to "." because check.names=TRUE
		# in the read.delim() function.
		normExprs <- read.delim(normExprsFile, stringsAsFactors=FALSE)
		rownames(normExprs) <- normExprs[,1]
		normExprs[,1] <- NULL

		# Make vectors of assay accessions.
		# Assay groups come in like e.g.:
		# 	assay 1<SEP>assay 2<SEP>assay 3
		# Split on "<SEP>" -- we use this as a separator because it is very
		# unlikely to crop up in actual assay names.
		# Use make.names to convert reserved characters to "." so that they
		# match the column headings from reading in the normalized expressions
		# above. 
		refAssays <- make.names(unlist(strsplit(refAssays, "<SEP>")))
		testAssays <- make.names(unlist(strsplit(testAssays, "<SEP>")))


		# Provisions for 2-colour designs:
		#	- If aValuesFile is not NULL, this means it's a 2-colour design.
		#	- Create an MAList object out of the M-values in normExprs and
		#	A-values in aValues. This is essentially equivalent to the
		#	ExpressionSet created for Affymetrix data.
		#	- Fit LM, get DE stats. Contrast is implied, do not need to specify contrasts matrix.
		#	- Write results and data for MvA plot -- think this part could be
		#	the same as for Affymetrix data?
		if(!is.null(aValuesFile)) {

			# Sort ref and test assays so they are in the same order.
			refAssays <- sort(refAssays)
			testAssays <- sort(testAssays)
			
			# Read in A-values
			aValues <- read.delim(aValuesFile, stringsAsFactors=FALSE)
			rownames(aValues) <- aValues[,1]
			aValues[,1] <- NULL

			# Strip off .Cy* endings to check the assay names are the same and
			# to subset columns from normExpr and aValues.
			refAssaysNoCy <- gsub(".Cy\\d$", "", refAssays)
			testAssaysNoCy <- gsub(".Cy\\d$", "", testAssays)

			# [double-]check that we have the same assay names for test and ref
			# without .Cy* endings.
			# Compare sorted values at each index position using
			# any(). If there are any mismatches this returns TRUE.
			if(any(refAssaysNoCy != testAssaysNoCy)) {
				stop("Differing assay names found in test and reference assay groups after removing \".Cy3\" and \".Cy5\". Please verify this experiment has a two-colour design.")
			}
			
			
			# Create "targets" dataframe for design matrix.
			# Vector of ref and test for Cy3-labelled samples.
			cy3 <- rep(NA, length(refAssays))
			cy3[grep(".Cy3$", refAssays)] <- "ref"
			cy3[grep(".Cy3$", testAssays)] <- "test"
			# Vector of ref and test for Cy5-labelled samples.
			cy5 <- rep(NA, length(refAssays))
			cy5[grep(".Cy5$", refAssays)] <- "ref"
			cy5[grep(".Cy5$", testAssays)] <- "test"
			# Check in case any of cy3 or cy5 are still NA -- something's wrong if so.
			if(any(is.na(cy3)) || any(is.na(cy5))) {
				stop("Some assays do not have an accession ending with \".Cy3\" or \".Cy5\". Please verify this experiment has a two-colour design.")
			}
			# Create "targets" dataframe like:
			#	-----------
			#		Cy3	Cy5
			#	A1	ref	test
			#	A2	test	ref
			#	A3	ref	test
			#	A4	test	ref
			#	-----------
			targets <- data.frame(Cy3=cy3, Cy5=cy5)
			# Rownames are assay names and should be the same as the column
			# headings in dataframes of M-values and A-values.
			rownames(targets) <- refAssaysNoCy

			# Design matrix
			designMatrix <- modelMatrix(targets, ref="ref")


			# Subset M-values and A-values
			mValuesForContrast <- normExprs[,refAssaysNoCy]
			aValuesForContrast <- aValues[,refAssaysNoCy]

			# Verify that rownames (i.e. design element IDs) are the same (and
			# in the same order) for subsetted M-values and A-values.
			if(any(rownames(mValuesForContrast) != rownames(aValuesForContrast))) {
				stop("Different design element IDs found in M-values and A-values. Please verify they are from the same experiment.")
			}

			# Create MAList object.
			# First need to make a normal list with "genes" (design element
			# IDs), "M", and "A" elements.
			maList <- list(genes=rownames(mValuesForContrast), M=mValuesForContrast, A=aValuesForContrast)
			# Make it into an "MAList-class" object
			maList <- new("MAList", maList)

			fit <- lmFit(maList, designMatrix)
	
		} # if(!is.null(aValuesFile))
		
		# Otherwise this is single colour Affymetrix array, make ExpressionSet and apply linear model fit.
		else {
		
			# Subset the required columns of the expressions matrix.
			exprsForContrast <- normExprs[,c(refAssays, testAssays)]
			
			# Make an ExpressionSet for this contrast.
			esetForContrast <- makeEset(exprsForContrast, refAssays, testAssays)

			# Make design matrix.
			designMatrix <- model.matrix(as.formula("~0 + groups"), data=esetForContrast)
			

			# Fit linear model.
			fit <- lmFit(esetForContrast, design=designMatrix)
			
			# simple contrast matrix.
			contrast <- as.matrix(c(-1,1))

			# Apply contrast.
			fit <- contrasts.fit(fit, contrasts=contrast)
		}

		# moderated t-stats and associated p-values, B-stats (log-odds).
		fitEbayes <- eBayes(fit)
		
		# adjust p-values for multiple testing using Benjamini and Hochberg FDR
		# correction, and put them into the object.
		fitEbayes$adjPvals <- rep(NA, length(fitEbayes$p.value))
		nonNaPvIndices <- which(!is.na(fitEbayes$p.value))
		fitEbayes$adjPvals[nonNaPvIndices] <- p.adjust(fitEbayes$p.value[nonNaPvIndices], method="fdr")
		
		# Relevant results to data frames.
		# For 2-colour: 		
		if(!is.null(aValuesFile)) {
			
			# Have to specify which column of fitEbayes$t and
			# fitEbayes$coefficients we want. This is always the first column.

			# results for heatmap matrix display:
			contrastResults <- data.frame(designElements=fitEbayes$genes, adjPval=fitEbayes$adjPvals, t=fitEbayes$t[,1], logFC=fitEbayes$coefficients[,1])
			# results to be used for MvA plot creation (as above but with average intensities and without t-stats:
			plotData <- data.frame(designElements=fitEbayes$genes, adjPval=fitEbayes$adjPvals, logFC=fitEbayes$coefficients[,1], avgExpr=fitEbayes$Amean)
		}
		# For Affymetrix 1-colour:
		else {
			# results for heatmap matrix display:
			contrastResults <- data.frame(designElements = featureNames(esetForContrast), adjPval = fitEbayes$adjPvals, t = fitEbayes$t, logFC = fitEbayes$coefficients)
			# results to be used for MvA plot creation (as above but with average intensities and without t-stats):
			plotData <- data.frame(designElements = featureNames(esetForContrast), adjPval = fitEbayes$adjPvals, logFC = fitEbayes$coefficients, avgExpr = fitEbayes$Amean)
		}

		# write results.
		write.table(contrastResults, file=resFile, row.names=FALSE, quote=FALSE, sep="\t")
		write.table(plotData, file=plotDataFile, row.names=FALSE, quote=FALSE, sep="\t")
		
	})
}




# makeEset()
# 	- Create a Biobase ExpressionSet out of the assay, contrast and expression data.
#
makeEset <<- function(exprsForContrast, refAssays, testAssays) {

	# load Biobase for building ExpressionSet
	library(Biobase)

	# Expression data
	exprsForContrast <- as.matrix(exprsForContrast)
	expressionData <- assayDataNew(storage.mode = "lockedEnvironment", exprs=exprsForContrast)
	featureNames(expressionData) <- rownames(exprsForContrast)
	sampleNames(expressionData) <- colnames(exprsForContrast)

	# Factor data
	groups <- factor(c(rep("ref", length(refAssays)), rep("test", length(testAssays))))
	groupsDF <- data.frame(groups=groups)
	rownames(groupsDF) <- c(refAssays, testAssays)
	groupsDF <- new("AnnotatedDataFrame", data=groupsDF)

	# Feature data (just design element IDs, dies without this)
	featureData <- new("AnnotatedDataFrame", data=data.frame(designElements=rownames(exprsForContrast)))
	featureNames(featureData) <- rownames(exprsForContrast)
	
	# Return an ExpressionSet with all this data
	return(new("ExpressionSet", assayData=expressionData, phenoData=groupsDF, featureData=featureData))
}




# Run with arguments if there are any, otherwise don't do anything.
args <- commandArgs(TRUE)
if(length(args) > 0) {
	do.call(diffAtlas_DE_limma, as.list(args))
}
